File: backend/src/caad_erp/bll/transactions.py
"""Transaction workflows coordinating validation and persistence.

The module defines command dataclasses that describe user intent for sales,
restocks, write-offs, credit payments, and inventory snapshots. High-level
functions transform those commands into normalized transaction rows, enforce
business rules through centralized validators, and maintain cache coherence so
reporting modules observe consistent state.
"""

import dataclasses
import datetime
import logging
import typing as t

from caad_erp import constants, dal
from caad_erp.bll import rules, runtime

logger = logging.getLogger(__name__)


@dataclasses.dataclass(frozen=True)
class SaleCommand:
    """User intent for creating a ``SALE`` transaction."""

    product_id: str
    salesman_id: str
    quantity: int
    total_revenue: int
    payment_type: constants.PaymentType
    notes: str | None = None


@dataclasses.dataclass(frozen=True)
class RestockCommand:
    """User intent for creating a ``RESTOCK`` transaction."""

    product_id: str
    salesman_id: str
    quantity: int
    total_cost: int
    notes: str | None = None


@dataclasses.dataclass(frozen=True)
class WriteOffCommand:
    """User intent for creating a ``WRITE_OFF`` transaction."""

    product_id: str
    salesman_id: str
    quantity: int
    notes: str | None = None


@dataclasses.dataclass(frozen=True)
class CreditPaymentCommand:
    """User intent for logging a ``CREDIT_PAYMENT`` transaction."""

    linked_transaction_id: str
    salesman_id: str
    total_revenue: int
    payment_type: constants.PaymentType
    notes: str | None = None


@dataclasses.dataclass(frozen=True)
class OpenStockCommand:
    """Instruction for creating an ``OPEN_STOCK`` transaction during archiving."""

    product_id: str
    salesman_id: str
    quantity: int
    total_revenue: int


@dataclasses.dataclass(frozen=True)
class VoidCommand:
    """User intent for voiding a prior transaction."""

    linked_transaction_id: str
    notes: str | None = None


# Rationale (Single Sale Workflow):
# 1. Soft-Delete Protection (DEVELOPER_GUIDE.md #Soft-delete):
#    - Inactive products and salesmen cannot participate in new sales so they instantly vanish
#      from POS views and active catalog selections.
#    - However, historical catalog rows remain in Products/Salesmen sheets to ensure past
#      TransactionLog rows are never orphaned or corrupted.
# 2. Flexible Discounts (DEVELOPER_GUIDE.md #Discounts):
#    - Revenue is not hard-coded to suggested Product.SellPrice. Custom prices and discounts
#      are handled by allowing any total_revenue >= 0.
# 3. Stock Availability Enforcement (DEVELOPER_GUIDE.md #Stock Availability Enforcement):
#    - SUM(QuantityChange) derives real-time inventory. Sales reduce stock and cannot exceed
#      currently available inventory balances.
RECORD_SALE_RULES: list[rules.BaseRule] = [
    rules.PRODUCT_EXISTS,
    rules.PRODUCT_IS_ACTIVE,
    rules.SALESMAN_EXISTS,
    rules.SALESMAN_IS_ACTIVE,
    rules.POSITIVE_QUANTITY,
    rules.NONNEGATIVE_REVENUE,
    rules.SUFFICIENT_STOCK,
]

# Rationale (Restock Workflow):
# 1. Active Catalog Attribution (DEVELOPER_GUIDE.md #RESTOCK):
#    - Inventory restocks increase physical stock (> 0) and must be attributed to an active product
#      and active salesman.
# 2. Donated & Promotional Items (DEVELOPER_GUIDE.md #RESTOCK / #Separate Revenue and Cost Columns):
#    - Total cost is stored as a negative delta in TotalCost.
#    - Total cost of 0 is explicitly allowed to support donated items, promotional samples,
#      or free inventory additions without breaking cost accounting.
RECORD_RESTOCK_RULES: list[rules.BaseRule] = [
    rules.PRODUCT_EXISTS,
    rules.PRODUCT_IS_ACTIVE,
    rules.SALESMAN_EXISTS,
    rules.SALESMAN_IS_ACTIVE,
    rules.POSITIVE_QUANTITY,
    rules.NONNEGATIVE_COST,
]

# Rationale (Write-Off Workflow):
# 1. Spoilage & Loss Tracking (DEVELOPER_GUIDE.md #WRITE_OFF):
#    - Write-offs represent inventory loss (spoilage, theft, loss, damage) without revenue or cost deltas.
# 2. Stock Balance Guard (DEVELOPER_GUIDE.md #Stock Availability Enforcement):
#    - Write-off quantities must be positive and cannot exceed currently available inventory.
RECORD_WRITE_OFF_RULES: list[rules.BaseRule] = [
    rules.PRODUCT_EXISTS,
    rules.PRODUCT_IS_ACTIVE,
    rules.SALESMAN_EXISTS,
    rules.SALESMAN_IS_ACTIVE,
    rules.POSITIVE_QUANTITY,
    rules.SUFFICIENT_WRITE_OFF_STOCK,
]

# Rationale (Credit Payment Workflow):
# 1. Credit Sale Link (DEVELOPER_GUIDE.md #Sell on Credit & Flexible Credit Payments):
#    - Must reference a valid SALE transaction recorded with PaymentType='OnCredit'.
# 2. Void Immutability Guard (DEVELOPER_GUIDE.md #Error Correction & Voiding Credit Payments):
#    - Voided sales cannot accept credit payments to prevent phantom debt settlements and keep audit logs immutable.
# 3. Flexible Payments & Overpayments (DEVELOPER_GUIDE.md #Overpayments & Interest):
#    - Payment amount must be strictly > 0. Payments are intentionally NOT capped by remaining debt,
#      allowing salesmen to record partial payments, full settlements, interest, or late fees.
# 4. Active Salesman:
#    - Cash collection must be logged by an active salesman.
RECORD_CREDIT_PAYMENT_RULES: list[rules.BaseRule] = [
    rules.CREDIT_SALE_LINK_ELIGIBLE,
    rules.POSITIVE_REVENUE,
    rules.SALESMAN_EXISTS,
    rules.SALESMAN_IS_ACTIVE,
]

# Rationale (Open Stock Workflow):
# 1. Period Initialization (DEVELOPER_GUIDE.md #OPEN_STOCK / #Archive Script):
#    - OPEN_STOCK entries seed beginning-of-period inventory during archive processing for active catalog items.
RECORD_OPEN_STOCK_RULES: list[rules.BaseRule] = [
    rules.PRODUCT_EXISTS,
    rules.PRODUCT_IS_ACTIVE,
    rules.SALESMAN_EXISTS,
    rules.SALESMAN_IS_ACTIVE,
    rules.POSITIVE_QUANTITY,
    rules.NONNEGATIVE_REVENUE,
]


def _ensure_transactions_cache(context: runtime.RuntimeContext) -> dict[str, t.Any]:
    bucket = runtime.get_cache_bucket(context, "transactions")
    if "all" not in bucket:
        all_transactions = list(dal.iter_transactions(context.workbook))
        bucket["all"] = all_transactions
        bucket["by_id"] = {
            transaction.transaction_id: transaction for transaction in all_transactions
        }
        logger.debug(
            "Populated transactions cache with %d entries",
            len(all_transactions),
        )
    return bucket


def _generate_transaction_id(when: datetime.datetime) -> str:
    return when.strftime("%Y%m%d%H%M%S%f")


def list_transactions(context: runtime.RuntimeContext) -> list[dal.TransactionRow]:
    cache = _ensure_transactions_cache(context)
    return list(cache["all"])


def get_transaction(
    context: runtime.RuntimeContext, transaction_id: str
) -> dal.TransactionRow:
    cache = _ensure_transactions_cache(context)
    try:
        return cache["by_id"][transaction_id]
    except KeyError as exc:
        logger.warning("Transaction lookup failed for id '%s'", transaction_id)
        raise rules.TransactionNotFoundError(
            f"[Transaction Exists] Unknown transaction id: {transaction_id}"
        ) from exc


def record_sale(
    context: runtime.RuntimeContext, command: SaleCommand
) -> dal.TransactionRow:
    """Validate and append a ``SALE`` transaction to the logger."""
    now = datetime.datetime.now(datetime.UTC)
    rules.enforce_rules(context, command, RECORD_SALE_RULES)

    transaction_id = _generate_transaction_id(when=now)
    transaction = _build_sale_transaction(
        command, transaction_id=transaction_id, timestamp=now
    )
    dal.append_transaction(context.workbook, transaction)
    runtime.invalidate_cache(context, "transactions")
    logger.info(
        "Recorded SALE transaction '%s' for product '%s' (quantity=%s, revenue=%s)",
        transaction.transaction_id,
        command.product_id,
        command.quantity,
        command.total_revenue,
    )
    return transaction


def record_bulk_sale(
    context: runtime.RuntimeContext, commands: list[SaleCommand]
) -> list[dal.TransactionRow]:
    """Validate and append a list of ``SALE`` transactions atomically."""
    rules.enforce_rules(context, commands, [rules.NON_EMPTY_BULK_SALE])

    for command in commands:
        rules.enforce_rules(
            context, command, RECORD_SALE_RULES[:-1]
        )  # validate individual item checks

    rules.enforce_rules(context, commands, [rules.BULK_SUFFICIENT_STOCK])

    recorded: list[dal.TransactionRow] = []
    for command in commands:
        recorded.append(record_sale(context, command))

    return recorded


def _build_sale_transaction(
    command: SaleCommand, *, transaction_id: str, timestamp: datetime.datetime
) -> dal.TransactionRow:
    quantity_change = -abs(command.quantity)
    return dal.TransactionRow(
        transaction_id=transaction_id,
        timestamp_iso=timestamp.isoformat(),
        transaction_type=constants.TransactionType.SALE.value,
        product_id=command.product_id,
        salesman_id=command.salesman_id,
        payment_type=command.payment_type.value,
        quantity_change=quantity_change,
        total_revenue=command.total_revenue,
        total_cost=0,
        linked_transaction_id=None,
        notes=command.notes,
    )


def record_restock(
    context: runtime.RuntimeContext, command: RestockCommand
) -> dal.TransactionRow:
    """Validate and append a ``RESTOCK`` transaction."""
    now = datetime.datetime.now(datetime.UTC)
    rules.enforce_rules(context, command, RECORD_RESTOCK_RULES)

    transaction_id = _generate_transaction_id(when=now)
    transaction = _build_restock_transaction(
        command, transaction_id=transaction_id, timestamp=now
    )
    dal.append_transaction(context.workbook, transaction)
    runtime.invalidate_cache(context, "transactions")
    logger.info(
        "Recorded RESTOCK transaction '%s' for product '%s' (quantity=%s, cost=%s)",
        transaction.transaction_id,
        command.product_id,
        command.quantity,
        command.total_cost,
    )
    return transaction


def _build_restock_transaction(
    command: RestockCommand, *, transaction_id: str, timestamp: datetime.datetime
) -> dal.TransactionRow:
    quantity_change = abs(command.quantity)
    cost_value = -abs(command.total_cost)
    return dal.TransactionRow(
        transaction_id=transaction_id,
        timestamp_iso=timestamp.isoformat(),
        transaction_type=constants.TransactionType.RESTOCK.value,
        product_id=command.product_id,
        salesman_id=command.salesman_id,
        payment_type=None,
        quantity_change=quantity_change,
        total_revenue=0,
        total_cost=cost_value,
        linked_transaction_id=None,
        notes=command.notes,
    )


def record_write_off(
    context: runtime.RuntimeContext, command: WriteOffCommand
) -> dal.TransactionRow:
    """Validate and append a ``WRITE_OFF`` transaction."""
    now = datetime.datetime.now(datetime.UTC)
    rules.enforce_rules(context, command, RECORD_WRITE_OFF_RULES)

    transaction_id = _generate_transaction_id(when=now)
    transaction = _build_write_off_transaction(
        command, transaction_id=transaction_id, timestamp=now
    )
    dal.append_transaction(context.workbook, transaction)
    runtime.invalidate_cache(context, "transactions")
    logger.info(
        "Recorded WRITE_OFF transaction '%s' for product '%s' (quantity=%s)",
        transaction.transaction_id,
        command.product_id,
        command.quantity,
    )
    return transaction


def _build_write_off_transaction(
    command: WriteOffCommand, *, transaction_id: str, timestamp: datetime.datetime
) -> dal.TransactionRow:
    quantity_change = -abs(command.quantity)
    return dal.TransactionRow(
        transaction_id=transaction_id,
        timestamp_iso=timestamp.isoformat(),
        transaction_type=constants.TransactionType.WRITE_OFF.value,
        product_id=command.product_id,
        salesman_id=command.salesman_id,
        payment_type=None,
        quantity_change=quantity_change,
        total_revenue=0,
        total_cost=0,
        linked_transaction_id=None,
        notes=command.notes,
    )


def record_credit_payment(
    context: runtime.RuntimeContext, command: CreditPaymentCommand
) -> dal.TransactionRow:
    """Append a ``CREDIT_PAYMENT`` transaction linked to an outstanding sale."""
    now = datetime.datetime.now(datetime.UTC)
    rules.enforce_rules(context, command, RECORD_CREDIT_PAYMENT_RULES)
    linked_sale = get_transaction(context, command.linked_transaction_id)

    transaction_id = _generate_transaction_id(when=now)
    transaction = _build_credit_payment_transaction(
        command,
        transaction_id=transaction_id,
        timestamp=now,
        product_id=linked_sale.product_id,
    )
    dal.append_transaction(context.workbook, transaction)
    runtime.invalidate_cache(context, "transactions")
    logger.info(
        "Recorded CREDIT_PAYMENT '%s' linked to '%s' (amount=%s)",
        transaction.transaction_id,
        command.linked_transaction_id,
        command.total_revenue,
    )
    return transaction


def _build_credit_payment_transaction(
    command: CreditPaymentCommand,
    *,
    transaction_id: str,
    timestamp: datetime.datetime,
    product_id: str,
) -> dal.TransactionRow:
    return dal.TransactionRow(
        transaction_id=transaction_id,
        timestamp_iso=timestamp.isoformat(),
        transaction_type=constants.TransactionType.CREDIT_PAYMENT.value,
        product_id=product_id,
        salesman_id=command.salesman_id,
        payment_type=command.payment_type.value,
        quantity_change=0,
        total_revenue=command.total_revenue,
        total_cost=0,
        linked_transaction_id=command.linked_transaction_id,
        notes=command.notes,
    )


def record_open_stock(
    context: runtime.RuntimeContext, command: OpenStockCommand
) -> dal.TransactionRow:
    """Append an ``OPEN_STOCK`` transaction for period initialization."""
    now = datetime.datetime.now(datetime.UTC)
    rules.enforce_rules(context, command, RECORD_OPEN_STOCK_RULES)

    transaction_id = _generate_transaction_id(when=now)
    transaction = _build_open_stock_transaction(
        command, transaction_id=transaction_id, timestamp=now
    )
    dal.append_transaction(context.workbook, transaction)
    runtime.invalidate_cache(context, "transactions")
    logger.info(
        "Recorded OPEN_STOCK transaction '%s' for product '%s' (quantity=%s, value=%s)",
        transaction.transaction_id,
        command.product_id,
        command.quantity,
        command.total_revenue,
    )
    return transaction


def _build_open_stock_transaction(
    command: OpenStockCommand, *, transaction_id: str, timestamp: datetime.datetime
) -> dal.TransactionRow:
    quantity_change = abs(command.quantity)
    return dal.TransactionRow(
        transaction_id=transaction_id,
        timestamp_iso=timestamp.isoformat(),
        transaction_type=constants.TransactionType.OPEN_STOCK.value,
        product_id=command.product_id,
        salesman_id=command.salesman_id,
        payment_type=None,
        quantity_change=quantity_change,
        total_revenue=command.total_revenue,
        total_cost=0,
        linked_transaction_id=None,
        notes=None,
    )


def record_void(
    context: runtime.RuntimeContext, command: VoidCommand
) -> dal.TransactionRow:
    """Record a ``VOID`` reversal negating a prior transaction."""
    now = datetime.datetime.now(datetime.UTC)
    logger.info("Recording VOID for transaction '%s'", command.linked_transaction_id)
    target = get_transaction(context, command.linked_transaction_id)
    rules.enforce_rules(context, target, [rules.VOID_TARGET_ELIGIBLE])

    reversal = _build_void_transaction(target, timestamp=now, notes=command.notes)
    dal.append_transaction(context.workbook, reversal)
    runtime.invalidate_cache(context, "transactions")
    logger.info(
        "Recorded VOID reversal '%s' for transaction '%s'",
        reversal.transaction_id,
        target.transaction_id,
    )
    return reversal


def _build_void_transaction(
    transaction: dal.TransactionRow,
    *,
    timestamp: datetime.datetime,
    notes: str | None,
) -> dal.TransactionRow:
    return dal.TransactionRow(
        transaction_id=_generate_transaction_id(when=timestamp),
        timestamp_iso=timestamp.isoformat(),
        transaction_type=constants.TransactionType.VOID.value,
        product_id=transaction.product_id,
        salesman_id=transaction.salesman_id,
        payment_type=transaction.payment_type,
        quantity_change=-transaction.quantity_change,
        total_revenue=-transaction.total_revenue,
        total_cost=-transaction.total_cost,
        linked_transaction_id=transaction.transaction_id,
        notes=notes,
    )

File: backend/src/bll/transactions.ts
/**
 * Transaction ledger domain handlers coordinating validation, business rules, and persistence.
 *
 * Implements high-level transaction ledger workflows (`listTransactions`, `getTransaction`,
 * `recordSale`, `recordBulkSale`, `recordRestock`, `recordWriteOff`, `recordCreditPayment`,
 * `recordOpenStock`, `recordVoid`) by colocating Zod command schemas, domain invariant rules,
 * and Data Access Layer (DAL) execution.
 */

import { z } from "zod"
import type { DB, TransactionRow } from "../dal/index.js"
import * as dal from "../dal/index.js"
import { paymentTypeValues } from "../dal/index.js"
import {
    EmptyBulkOperationError,
    IneligibleCreditSaleError,
    IneligibleVoidTargetError,
    InsufficientStockError,
    InvalidAttributeError,
    InvalidMonetaryValueError,
    InvalidQuantityError,
    ProductInactiveError,
    SalesmanInactiveError,
    TransactionNotFoundError,
} from "./errors.js"
import { v7 as uuidv7 } from "uuid"
import { getProduct } from "./products.js"
import { calculateInventory } from "./reports.js"
import { getSalesman } from "./salesmen.js"
import { validateSchema } from "./validator.js"

/**
 * Generates a RFC 9562 compliant UUID v7 transaction identifier string.
 *
 * @returns Time-ordered 36-character UUID v7 string.
 */
export function generateTransactionId(): string {
    return uuidv7()
}

/**
 * Retrieves all transaction records from the ledger log.
 *
 * @param db - Active database client instance.
 * @returns Array of all {@link TransactionRow} items.
 */
export function listTransactions(db: DB): TransactionRow[] {
    return dal.listTransactions(db)
}

/**
 * Zod validation schema and command payload for retrieving a transaction by identifier.
 */
export const getTransactionSchema = z.object({
    id: z
        .string()
        .trim()
        .refine((val) => val.length >= 1, {
            message: "Transaction ID must be provided",
            params: { errorClass: InvalidAttributeError },
        }),
})

export type GetTransactionCommand = z.infer<typeof getTransactionSchema>

/**
 * Retrieves a single transaction record by identifier.
 *
 * @param db - Active database client instance.
 * @param id - Unique transaction identifier.
 * @returns The matching {@link TransactionRow}.
 * @throws {@link InvalidAttributeError} If transaction ID is empty string.
 * @throws {@link TransactionNotFoundError} If no transaction exists with the given ID.
 */
export function getTransaction(db: DB, id: string): TransactionRow {
    validateSchema(getTransactionSchema, { id })
    const all = dal.listTransactions(db)
    const transaction = all.find((tx) => tx.id === id)
    if (!transaction) {
        throw new TransactionNotFoundError(`Unknown transaction id: ${id}`)
    }
    return transaction
}

/**
 * Zod validation schema and command payload for recording a sale transaction.
 */
export const saleCommandSchema = z.object({
    productId: z
        .string()
        .trim()
        .refine((val) => val.length >= 1, {
            message: "Product ID must be provided",
            params: { errorClass: InvalidAttributeError },
        }),
    salesmanId: z
        .string()
        .trim()
        .refine((val) => val.length >= 1, {
            message: "Salesman ID must be provided",
            params: { errorClass: InvalidAttributeError },
        }),
    quantity: z
        .number()
        .refine((val) => Number.isInteger(val), {
            message: "Quantity must be an integer",
            params: { errorClass: InvalidQuantityError },
        })
        .refine((val) => val > 0, {
            message: "Quantity must be greater than zero",
            params: { errorClass: InvalidQuantityError },
        }),
    totalRevenue: z
        .number()
        .refine((val) => Number.isInteger(val), {
            message: "Amount must be an integer",
            params: { errorClass: InvalidMonetaryValueError },
        })
        .refine((val) => val >= 0, {
            message: "Amount must be zero or positive",
            params: { errorClass: InvalidMonetaryValueError },
        }),
    paymentType: z.enum(paymentTypeValues),
    notes: z.string().nullable().optional(),
})

export type SaleCommand = z.infer<typeof saleCommandSchema>

/**
 * Validates input payload and records a `SALE` transaction in the ledger.
 *
 * @param db - Active database client instance.
 * @param command - Raw input object matching {@link SaleCommand}.
 * @returns The newly recorded {@link TransactionRow}.
 * @throws {@link ProductNotFoundError} If product ID does not exist.
 * @throws {@link ProductInactiveError} If product is inactive.
 * @throws {@link SalesmanNotFoundError} If salesman ID does not exist.
 * @throws {@link SalesmanInactiveError} If salesman is inactive.
 * @throws {@link InsufficientStockError} If requested quantity exceeds available stock.
 */
export function recordSale(db: DB, command: SaleCommand): TransactionRow {
    // Validate input payload structure and boundary types using Zod
    const validated = validateSchema(saleCommandSchema, command)

    // Enforce domain rule checking that referenced product exists and is active
    const product = getProduct(db, validated.productId)
    if (!product.isActive) {
        throw new ProductInactiveError(`Product '${validated.productId}' is inactive`)
    }

    // Enforce domain rule checking that referenced salesman exists and is active
    const salesman = getSalesman(db, validated.salesmanId)
    if (!salesman.isActive) {
        throw new SalesmanInactiveError(`Salesman '${validated.salesmanId}' is inactive`)
    }

    // Enforce domain rule checking sufficient available inventory stock
    const inventory = calculateInventory(db)
    const availableStock = inventory[validated.productId] ?? 0
    if (validated.quantity > availableStock) {
        throw new InsufficientStockError(
            `Insufficient stock for product '${validated.productId}': available ${availableStock}, requested ${validated.quantity}`,
        )
    }

    // Build and persist normalized SALE transaction record
    const now = new Date()
    const transactionRecord: TransactionRow = {
        id: generateTransactionId(),
        timestampIso: now.toISOString(),
        transactionType: "SALE",
        productId: validated.productId,
        salesmanId: validated.salesmanId,
        paymentType: validated.paymentType,
        quantityChange: -Math.abs(validated.quantity),
        totalRevenue: validated.totalRevenue,
        totalCost: 0,
        linkedTransactionId: null,
        notes: validated.notes ?? null,
    }

    return dal.appendTransaction(db, transactionRecord)
}

/**
 * Validates and records a batch list of `SALE` transactions atomically with cart stock checking.
 *
 * Implements an all-or-nothing checkout workflow that validates cart item availability
 * prior to appending sale transactions to the ledger log in sequence.
 *
 * @param db - Active database client instance.
 * @param commands - List of raw sale command objects matching {@link SaleCommand}.
 * @returns Array of recorded {@link TransactionRow} items.
 * @throws {@link EmptyBulkOperationError} If commands list is empty.
 * @throws {@link InsufficientStockError} If aggregate cart quantity exceeds available stock.
 */
export function recordBulkSale(db: DB, commands: SaleCommand[]): TransactionRow[] {
    // Ensure cart contains at least one item
    if (!commands || commands.length === 0) {
        throw new EmptyBulkOperationError("Bulk sale requires at least one item")
    }

    // Validate payload schemas and sum total requested quantity per product ID
    const aggregateQuantities: Record<string, number> = {}
    for (const cmd of commands) {
        const validated = validateSchema(saleCommandSchema, cmd)
        aggregateQuantities[validated.productId] =
            (aggregateQuantities[validated.productId] ?? 0) + validated.quantity
    }

    // Ensure aggregate cart quantities do not exceed available stock
    const inventory = calculateInventory(db)
    for (const [productId, requestedTotal] of Object.entries(aggregateQuantities)) {
        const availableStock = inventory[productId] ?? 0
        if (requestedTotal > availableStock) {
            throw new InsufficientStockError(
                `Insufficient stock for product '${productId}': available ${availableStock}, total requested in cart ${requestedTotal}`,
            )
        }
    }

    // Record each sale transaction in the ledger log
    const recorded: TransactionRow[] = []
    for (const cmd of commands) {
        recorded.push(recordSale(db, cmd))
    }
    return recorded
}

/**
 * Zod validation schema and command payload for recording a restock transaction.
 */
export const restockCommandSchema = z.object({
    productId: z
        .string()
        .trim()
        .refine((val) => val.length >= 1, {
            message: "Product ID must be provided",
            params: { errorClass: InvalidAttributeError },
        }),
    salesmanId: z
        .string()
        .trim()
        .refine((val) => val.length >= 1, {
            message: "Salesman ID must be provided",
            params: { errorClass: InvalidAttributeError },
        }),
    quantity: z
        .number()
        .refine((val) => Number.isInteger(val), {
            message: "Quantity must be an integer",
            params: { errorClass: InvalidQuantityError },
        })
        .refine((val) => val > 0, {
            message: "Quantity must be greater than zero",
            params: { errorClass: InvalidQuantityError },
        }),
    totalCost: z
        .number()
        .refine((val) => Number.isInteger(val), {
            message: "Amount must be an integer",
            params: { errorClass: InvalidMonetaryValueError },
        })
        .refine((val) => val >= 0, {
            message: "Amount must be zero or positive",
            params: { errorClass: InvalidMonetaryValueError },
        }),
    notes: z.string().nullable().optional(),
})

export type RestockCommand = z.infer<typeof restockCommandSchema>

/**
 * Validates input payload and records a `RESTOCK` transaction in the ledger.
 *
 * @param db - Active database client instance.
 * @param command - Raw input object matching {@link RestockCommand}.
 * @returns The newly recorded {@link TransactionRow}.
 * @throws {@link ProductNotFoundError} If product ID does not exist.
 * @throws {@link ProductInactiveError} If product is inactive.
 * @throws {@link SalesmanNotFoundError} If salesman ID does not exist.
 * @throws {@link SalesmanInactiveError} If salesman is inactive.
 */
export function recordRestock(db: DB, command: RestockCommand): TransactionRow {
    // Validate input payload structure and boundary types using Zod
    const validated = validateSchema(restockCommandSchema, command)

    // Enforce domain rule checking that referenced product exists and is active
    const product = getProduct(db, validated.productId)
    if (!product.isActive) {
        throw new ProductInactiveError(`Product '${validated.productId}' is inactive`)
    }

    // Enforce domain rule checking that referenced salesman exists and is active
    const salesman = getSalesman(db, validated.salesmanId)
    if (!salesman.isActive) {
        throw new SalesmanInactiveError(`Salesman '${validated.salesmanId}' is inactive`)
    }

    // Build and persist normalized RESTOCK transaction record
    const now = new Date()
    const transactionRecord: TransactionRow = {
        id: generateTransactionId(),
        timestampIso: now.toISOString(),
        transactionType: "RESTOCK",
        productId: validated.productId,
        salesmanId: validated.salesmanId,
        paymentType: null,
        quantityChange: Math.abs(validated.quantity),
        totalRevenue: 0,
        totalCost: -Math.abs(validated.totalCost),
        linkedTransactionId: null,
        notes: validated.notes ?? null,
    }

    return dal.appendTransaction(db, transactionRecord)
}

/**
 * Zod validation schema and command payload for recording a write-off transaction.
 */
export const writeOffCommandSchema = z.object({
    productId: z
        .string()
        .trim()
        .refine((val) => val.length >= 1, {
            message: "Product ID must be provided",
            params: { errorClass: InvalidAttributeError },
        }),
    salesmanId: z
        .string()
        .trim()
        .refine((val) => val.length >= 1, {
            message: "Salesman ID must be provided",
            params: { errorClass: InvalidAttributeError },
        }),
    quantity: z
        .number()
        .refine((val) => Number.isInteger(val), {
            message: "Quantity must be an integer",
            params: { errorClass: InvalidQuantityError },
        })
        .refine((val) => val > 0, {
            message: "Quantity must be greater than zero",
            params: { errorClass: InvalidQuantityError },
        }),
    notes: z.string().nullable().optional(),
})

export type WriteOffCommand = z.infer<typeof writeOffCommandSchema>

/**
 * Validates input payload and records a `WRITE_OFF` transaction in the ledger.
 *
 * @param db - Active database client instance.
 * @param command - Raw input object matching {@link WriteOffCommand}.
 * @returns The newly recorded {@link TransactionRow}.
 * @throws {@link ProductNotFoundError} If product ID does not exist.
 * @throws {@link ProductInactiveError} If product is inactive.
 * @throws {@link SalesmanNotFoundError} If salesman ID does not exist.
 * @throws {@link SalesmanInactiveError} If salesman is inactive.
 * @throws {@link InsufficientStockError} If write-off quantity exceeds available stock.
 */
export function recordWriteOff(db: DB, command: WriteOffCommand): TransactionRow {
    // Validate input payload structure and boundary types using Zod
    const validated = validateSchema(writeOffCommandSchema, command)

    // Enforce domain rule checking that referenced product exists and is active
    const product = getProduct(db, validated.productId)
    if (!product.isActive) {
        throw new ProductInactiveError(`Product '${validated.productId}' is inactive`)
    }

    // Enforce domain rule checking that referenced salesman exists and is active
    const salesman = getSalesman(db, validated.salesmanId)
    if (!salesman.isActive) {
        throw new SalesmanInactiveError(`Salesman '${validated.salesmanId}' is inactive`)
    }

    // Enforce domain rule checking available inventory stock for write-off
    const inventory = calculateInventory(db)
    const availableStock = inventory[validated.productId] ?? 0
    if (validated.quantity > availableStock) {
        throw new InsufficientStockError(
            `Cannot write off ${validated.quantity} units of product '${validated.productId}': only ${availableStock} available`,
        )
    }

    // Build and persist normalized WRITE_OFF transaction record
    const now = new Date()
    const transactionRecord: TransactionRow = {
        id: generateTransactionId(),
        timestampIso: now.toISOString(),
        transactionType: "WRITE_OFF",
        productId: validated.productId,
        salesmanId: validated.salesmanId,
        paymentType: null,
        quantityChange: -Math.abs(validated.quantity),
        totalRevenue: 0,
        totalCost: 0,
        linkedTransactionId: null,
        notes: validated.notes ?? null,
    }

    return dal.appendTransaction(db, transactionRecord)
}

/**
 * Zod validation schema and command payload for recording a credit payment.
 */
export const creditPaymentCommandSchema = z.object({
    linkedTransactionId: z
        .string()
        .trim()
        .refine((val) => val.length >= 1, {
            message: "Transaction ID must be provided",
            params: { errorClass: InvalidAttributeError },
        }),
    salesmanId: z
        .string()
        .trim()
        .refine((val) => val.length >= 1, {
            message: "Salesman ID must be provided",
            params: { errorClass: InvalidAttributeError },
        }),
    totalRevenue: z
        .number()
        .refine((val) => Number.isInteger(val), {
            message: "Payment amount must be an integer",
            params: { errorClass: InvalidMonetaryValueError },
        })
        .refine((val) => val > 0, {
            message: "Payment amount must be greater than zero",
            params: { errorClass: InvalidMonetaryValueError },
        }),
    paymentType: z.enum(paymentTypeValues),
    notes: z.string().nullable().optional(),
})

export type CreditPaymentCommand = z.infer<typeof creditPaymentCommandSchema>

/**
 * Validates input payload and records a `CREDIT_PAYMENT` transaction linked to an outstanding sale.
 *
 * @param db - Active database client instance.
 * @param command - Raw input object matching {@link CreditPaymentCommand}.
 * @returns The newly recorded {@link TransactionRow}.
 * @throws {@link TransactionNotFoundError} If linked transaction ID does not exist.
 * @throws {@link IneligibleCreditSaleError} If linked transaction is not an active credit sale.
 * @throws {@link SalesmanNotFoundError} If salesman ID does not exist.
 * @throws {@link SalesmanInactiveError} If salesman is inactive.
 */
export function recordCreditPayment(db: DB, command: CreditPaymentCommand): TransactionRow {
    // Validate input payload structure and boundary types using Zod
    const validated = validateSchema(creditPaymentCommandSchema, command)

    // Enforce domain rule checking that referenced salesman exists and is active
    const salesman = getSalesman(db, validated.salesmanId)
    if (!salesman.isActive) {
        throw new SalesmanInactiveError(`Salesman '${validated.salesmanId}' is inactive`)
    }

    // Enforce domain rule checking that linked transaction exists, is a SALE, and was recorded OnCredit
    const linkedSale = getTransaction(db, validated.linkedTransactionId)
    if (linkedSale.transactionType !== "SALE") {
        throw new IneligibleCreditSaleError("Credit payments must reference a SALE transaction")
    }
    if (linkedSale.paymentType !== "OnCredit") {
        throw new IneligibleCreditSaleError("Linked sale is not recorded as credit")
    }

    // Enforce domain rule checking that linked sale transaction is not voided
    const allTxs = dal.listTransactions(db)
    const voidedTxIds = new Set(
        allTxs
            .filter((tx) => tx.transactionType === "VOID" && tx.linkedTransactionId)
            .map((tx) => tx.linkedTransactionId as string),
    )
    if (voidedTxIds.has(linkedSale.id)) {
        throw new IneligibleCreditSaleError("Cannot process credit payment for voided transaction")
    }

    // Build and persist normalized CREDIT_PAYMENT transaction record
    const now = new Date()
    const transactionRecord: TransactionRow = {
        id: generateTransactionId(),
        timestampIso: now.toISOString(),
        transactionType: "CREDIT_PAYMENT",
        productId: linkedSale.productId,
        salesmanId: validated.salesmanId,
        paymentType: validated.paymentType,
        quantityChange: 0,
        totalRevenue: validated.totalRevenue,
        totalCost: 0,
        linkedTransactionId: validated.linkedTransactionId,
        notes: validated.notes ?? null,
    }

    return dal.appendTransaction(db, transactionRecord)
}

/**
 * Zod validation schema and command payload for voiding a transaction.
 */
export const voidCommandSchema = z.object({
    linkedTransactionId: z
        .string()
        .trim()
        .refine((val) => val.length >= 1, {
            message: "Transaction ID must be provided",
            params: { errorClass: InvalidAttributeError },
        }),
    notes: z.string().nullable().optional(),
})

export type VoidCommand = z.infer<typeof voidCommandSchema>

/**
 * Validates input payload and records a `VOID` transaction reversing a prior entry.
 *
 * @param db - Active database client instance.
 * @param command - Raw input object matching {@link VoidCommand}.
 * @returns The newly recorded reversing {@link TransactionRow}.
 * @throws {@link TransactionNotFoundError} If target transaction ID does not exist.
 * @throws {@link IneligibleVoidTargetError} If target transaction is already a VOID entry.
 */
export function recordVoid(db: DB, command: VoidCommand): TransactionRow {
    // Validate input payload structure using Zod
    const validated = validateSchema(voidCommandSchema, command)

    // Enforce domain rule checking that target transaction exists
    const target = getTransaction(db, validated.linkedTransactionId)

    // Enforce domain rule checking that target transaction is not already a VOID entry
    if (target.transactionType === "VOID") {
        throw new IneligibleVoidTargetError("Cannot void a VOID transaction")
    }

    // Build and persist exact reversing VOID transaction record
    const now = new Date()
    const reversalRecord: TransactionRow = {
        id: generateTransactionId(),
        timestampIso: now.toISOString(),
        transactionType: "VOID",
        productId: target.productId,
        salesmanId: target.salesmanId,
        paymentType: target.paymentType,
        quantityChange: -target.quantityChange,
        totalRevenue: -target.totalRevenue,
        totalCost: -target.totalCost,
        linkedTransactionId: target.id,
        notes: validated.notes ?? null,
    }

    return dal.appendTransaction(db, reversalRecord)
}
